pyiron_atomistics/sphinx/output_parser.py

```diff
@@ -386,3 +389,116 @@ def get_convergence(self):
     def get_fermi(self):
         pattern = r"Fermi energy:\s+(\d+\.\d+)\s+eV"
         return np.array(re.findall(pattern, self.log_main)).astype(float)
+
+
+class SphinxWavesParser:
+    """ Class to read SPHInX waves.sxb files (HDF5 format)
+    
+        Initialize with waves.sxb filename, or use load ()
+    
+    """
+    
+    def __init__(self, file_name="waves.sxb", cwd="."):
+        """
+        Args:
+            file_name (str): file name
+            cwd (str): directory path
+        """
+        if Path(file_name).is_absolute():
+            self.load(Path(file_name))
+        else:
+            path = Path(cwd) / Path(file_name)
+            self.load(path)
+        
+        
+
+    def load(self, filename):
+        """ Load a waves.sxb file (HDF5 format)
+        
+            filename: file name
+        """
+        self.wfile = h5py.File (filename)
+        self._eps = None
+
+    @property
+    def _n_gk(self):
+        return self.wfile['meshDim'][0]  
+    
+    @property
+    def _fft_idx(self):
+        fft_idx=[]
+        off=0
+        for ngk in self._n_gk:
+            fft_idx.append (self.wfile['fftIdx'][off:off+ngk])
+            off += ngk
+        return fft_idx
+    
+    @property
+    def mesh(self):
+        return self.wfile['meshDim'][:]
+
+    @property
+    def Nx(self):
+        return self.wfile['meshDim'][0]  
+
+    @property
+    def Ny(self):
+        return self.wfile['meshDim'][1]  
+
+    @property
+    def Nz(self):
+        return self.wfile['meshDim'][2]        
+        
+    @property
+    def n_states(self):
+        return self.wfile["nPerK"][0]
+
+    @property
+    def n_spin(self):
+        return self.wfile['nSpin'].shape[0]
+
+    @property
+    def k_weights(self):
+        return self.wfile['kWeights'][:]
+
+    @property
+    def k_vec(self):
+        return self.wfile['kVec'][:]
+
+    @property
+    def eps(self):
+        """All eigenvalues (in Hartree) as (nk,n_states) block"""
+        if (self._eps is None):
+            self._eps = self.wfile['eps'][:].reshape (-1,self.n_spin,self.n_states)
+        return self._eps.T #change
+
+    # Define as separate method and speed it up with numba
+    @staticmethod
+    @numba.jit
+    def _fillin(res,psire,psiim,fft_idx):
+        """Distributes condensed psi (real, imag) on full FFT mesh"""
+        rflat=res.flat
+        for ig in range(fft_idx.shape[0]):
+            rflat[fft_idx[ig]] = complex(psire[ig], psiim[ig])
```

Comment:
Is this really necessary to pull in `numba` here?  Afaik we don't use it elsewhere.

Could you try something like this

```python
res.flat[fft_idx] = psire +1j * psiim
```

Reply:
If this doesn't work directly numpy.put and friends may be helpful.

Reply:
This was @freyso 's call.

Reply:
In principle `res.flat[fft_idx] = psire +1j * psiim` might work.

Reply:
yes, this works at the expense of creating two extra copies of the data (namely first multiplying with 1j, which returns a complex vector where all real values are zero, and then adding psire (yielding a second vector), which is then fed into the FFT mesh. This might be fine here, but since I come from HPC, it never occurred to me to go this way, because the loop form doesn't need these superfluous vectors.
If we agree that performance doesn't matter, change it to the proposed version for the sake of readability.

Reply:
for loops tend to be orders of magnitude slower than `numpy` operations, so probably it's not only readable, but also faster.

Reply:
that's what the numba.jit is for. python loops are horribly slow, compiled loops are blazingly fast.

Reply:
ah ok I see, then I might even prefer to remove both numba and the for loop XD but you are right that we should probably first check the performance

Reply:
Agree with the performance concerns, but I'd really want to avoid pulling in extra dependencies for one method, especially something that couples tightly to numpy.  Making an extra copy will still be much faster than a python loop though.

Anyway, I found this snippet that supposedly makes a complex array without the extra copy [here](https://stackoverflow.com/questions/61356763/numpy-efficient-way-to-create-a-complex-array-from-two-real-arrays).

```np.stack((a, b), axis=-1).view(np.complex).squeeze(-1)```

I have not checked this.

Reply:
no need to check performance, the numpy version runs slower by at most a factor 3 in most situations, and that is acceptable because whatever you do next with the psi will be slower anyway. Probably the actual speed loss is less, as the indexed assignment is the actual bottleneck. As I said, it didn't occur to me to do it the numpy way, but I entirely agree that the numpy version is more readable, so kick out numba & the loop. pyiron is not about HPC.

Reply:
as for @pmrv 's version: while avoiding one of the two superfluous copies, it is far less readable.

Reply:
and then merge the reorder code into the read function, because the extra reorder function was only needed to encapsulate the loop for numba.jit

Reply:
Just did a quick test the `numba `JIT version takes 2.5 s and `numpy `version takes 0.1 s. So I will remove the `numba `implementation

Reply:
also, I stuck with this `res.flat[fft_idx] = psire +1j * psiim` as I felt this was readable to me.